libsql-sqlite3/src/vectorIndex.c

```diff
@@ -0,0 +1,462 @@
+  if( pVectorInRow->nKeys == 1 && sqlite3_value_type(pVectorInRow->pKeyValues + 0) == SQLITE_INTEGER ){
+    return sqlite3_value_int64(pVectorInRow->pKeyValues);
+  }
+  return 0;
+}
+
+int vectorInRowTryGetRowid(const VectorInRow *pVectorInRow, u64 *nRowid) {
+  if( pVectorInRow->nKeys != 1 ){
+    return -1;
+  }
+  if( sqlite3_value_type(vectorInRowKey(pVectorInRow, 0)) != SQLITE_INTEGER ){
+    return -1;
+  }
+  *nRowid = sqlite3_value_int64(vectorInRowKey(pVectorInRow, 0));
+  return 0;
+}
+
+int vectorInRowPlaceholderRender(const VectorInRow *pVectorInRow, char *pBuf, int nBufSize) {
+  int i;
+  assert( pVectorInRow->nKeys > 0 );
+  if( nBufSize < 2 * pVectorInRow->nKeys ){
+    return -1;
+  }
+  for(i = 0; i < pVectorInRow->nKeys; i++){
+    *(pBuf++) = '?';
+    *(pBuf++) = ',';
+  }
+  *(pBuf - 1) = '\0';
+  return 0;
+}
+
+int vectorInRowAlloc(sqlite3 *db, const UnpackedRecord *pRecord, VectorInRow *pVectorInRow, char **pzErrMsg) {
+  int rc = SQLITE_OK;
+  int type, dims;
+  struct sqlite3_value *pVectorValue = pRecord->aMem + 0;
+  pVectorInRow->pKeyValues = pRecord->aMem + 1;
+  pVectorInRow->nKeys = pRecord->nField - 1;
+  pVectorInRow->pVector = NULL;
+
+  if( pVectorInRow->nKeys <= 0 ){
+    rc = SQLITE_ERROR;
+    goto out;  
+  }
+
+  if( sqlite3_value_type(pVectorValue)==SQLITE_NULL ){
+    rc = SQLITE_OK;
+    goto out;
+  }
+
+  if( detectVectorParameters(pVectorValue, VECTOR_TYPE_FLOAT32, &type, &dims, pzErrMsg) != 0 ){
+    rc = SQLITE_ERROR;
+    goto out;
+  }
+
+  pVectorInRow->pVector = vectorAlloc(type, dims);
+  if( pVectorInRow->pVector == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+
+  if( sqlite3_value_type(pVectorValue) == SQLITE_BLOB ){
+    vectorInitFromBlob(pVectorInRow->pVector, sqlite3_value_blob(pVectorValue), sqlite3_value_bytes(pVectorValue));
+  } else if( sqlite3_value_type(pVectorValue) == SQLITE_TEXT ){ 
+    // users can put strings (e.g. '[1,2,3]') in the table and we should process them correctly
+    if( vectorParse(pVectorValue, pVectorInRow->pVector, pzErrMsg) != 0 ){
+      rc = SQLITE_ERROR;
+      goto out;
+    }
+  }
+  rc = SQLITE_OK;
+out:
+  if( rc != SQLITE_OK ){
+    vectorFree(pVectorInRow->pVector);
+  }
+  return rc;
+}
+
+void vectorInRowFree(sqlite3 *db, VectorInRow *pVectorInRow) {
+  vectorFree(pVectorInRow->pVector);
+}
+
+/**************************************************************************
+** VectorOutRows utilities
+****************************************************************************/
+
+int vectorOutRowsAlloc(sqlite3 *db, VectorOutRows *pRows, int nRows, int nCols, char firstColumnAff){
+  assert( nCols > 0 && nRows >= 0 );
+  pRows->nRows = nRows;
+  pRows->nCols = nCols;
+  pRows->aRowids = NULL;
+  pRows->ppValues = NULL;
+
+  if( (u64)nRows * (u64)nCols > VECTOR_OUT_ROWS_MAX_CELLS ){
+    return SQLITE_NOMEM_BKPT;
+  }
+
+  if( nCols == 1 && firstColumnAff == SQLITE_AFF_INTEGER ){
```

Comment:
Is this always true that when there's 1 integer column then it will be a rowid? What about the tables with no rowid that have single integer primary key?

Reply:
I must rename `aRowids` array to something more generic (`aIntColumn` maybe).

This code actually just wants to optimize use case when we need to return single integer column in the output.

Reply:
I understand now. Thanks